extensions/interactions/base_test.py

```diff
@@ -265,124 +265,74 @@ def test_html_field_types_to_rule_specs_mapping_are_valid(self):
+        # unique. After verification we delete the key for comparison with
+        # the generated html_field_types_to_rule_specs dict. We compare
+        # everything except the format, because the format can't be generated
+        # from the rule_descriptions.json file.
+        for html_type, html_type_dict in (
+                html_field_types_to_rule_specs_dict.items()):
+            self.assertTrue('format' in html_type_dict)
+            self.assertTrue(
+                html_type_dict['format'] in allowed_html_type_formats)
+            del html_type_dict['format']
 
         for interaction_id, interaction_rules in (
                 rule_descriptions_dict.items()):
             for rule_type, rule_description in interaction_rules.items():
                 description = rule_description['description']
+                # Extract the html_types from the rule description.
+                html_types_in_description = (
+                    re.findall(r'{{[a-z]\|([^}]*)}', description))
+                for html_type in html_types_in_description:
+                    if 'Html' in html_type and html_type not in html_types:
+                        html_types.append(html_type)
+
                 for html_type in html_types:
-                    if html_type['type'] in description:
-                        if (
-                                html_type['type'] not in
-                                generated_html_field_types_dict):
-                            generated_html_field_types_dict[
-                                html_type['type']] = {}
-                        generated_html_field_types_dict[
-                            html_type['type']]['interactionId'] = (
-                                interaction_id)
-                        generated_html_field_types_dict[html_type['type']][
-                            'format'] = html_type['format']
-                        if (
-                                'ruleTypes' not in
-                                generated_html_field_types_dict[
-                                    html_type['type']]):
-                            generated_html_field_types_dict[html_type['type']][
-                                'ruleTypes'] = {}
-                        generated_html_field_types_dict[html_type['type']][
-                            'ruleTypes'][rule_type] = {}
-                        generated_html_field_types_dict[html_type['type']][
-                            'ruleTypes'][rule_type]['htmlInputVariables'] = (
-                                ['x'])
-                        if '{{y|%s}}' % html_type['type'] in description:
-                            generated_html_field_types_dict[html_type['type']][
-                                'ruleTypes'][rule_type][
-                                    'htmlInputVariables'].append('y')
+                    if html_type in description:
+                        if html_type not in generated_html_field_types_dict:
```

Comment:
Make your generated_html_field_types_dict a collections.defaultdict, then you don't have to deal with all this stuff. Ditto elsewhere below. Otherwise it becomes very hard to read.

Try to keep your code as elegant, simple and readable as you can. Leave blank lines between logical sections.

Reply:
done